orttraining/orttraining/training_api/utils.cc

```diff
@@ -0,0 +1,78 @@
+#include "core/graph/model.h"
+#include "core/framework/session_state.h"
+#include "core/framework/ort_value.h"
+#include "core/framework/tensor.h"
+#include "core/framework/allocator.h"
+
+#include "orttraining/training_api/include/utils.h"
+
+namespace onnxruntime {
+namespace training {
+namespace api {
+namespace utils {
+
+// TODO: consolidate the gradient names with frontend tooling
+const std::vector<std::string> GRAD_SUFFIX{"_grad.accumulation.buffer", "_grad", "_grad.accumulation.out"};
+
+void GetGraphInputOutputNames(const std::unique_ptr<onnxruntime::InferenceSession>& session_object,
+                              std::vector<std::string>& input_names,
+                              std::vector<std::string>& output_names) {
+  auto get_names = [](const std::vector<const NodeArg*>* node_args, std::vector<std::string>& names) {
+    ORT_ENFORCE(nullptr != node_args);
+    for (const auto* arg : *node_args) {
+      names.push_back(arg->Name());
+    }
+  };
+  const auto& retval_input = session_object->GetModelInputs();
+  ORT_ENFORCE(retval_input.first.IsOK());
+  get_names(retval_input.second, input_names);
+  const auto& retval_output = session_object->GetModelOutputs();
+  ORT_ENFORCE(retval_output.first.IsOK());
+  get_names(retval_output.second, output_names);
+}
+
+bool GetParamNameFromSuffix(const std::string& name, const std::string& suffix, std::string& param_name) {
+  bool endswith = std::equal(suffix.rbegin(), suffix.rend(), name.rbegin());
+  if (endswith) {
+    param_name = name.substr(0, name.length() - suffix.length());
+    return true;
+  } else {
+    param_name = "";
+    return false;
+  }
+}
+
+bool GetParamNameFromGradient(const std::string& grad_name, std::string& param_name) {
+  for (auto& suffix : GRAD_SUFFIX) {
+    if (GetParamNameFromSuffix(grad_name, suffix, param_name)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+Status OrtValueLike(const SessionState& sess_state, const OrtValue& input_val, OrtValue& output_val) {
```

Comment:
Shall we call it CreateOrtValueFillingWithZero?  I mean this name is part of torch's naming style...

Reply:
I think it is okay